### DISCUSSION TITLE
Select to Functionality

### DISCUSSION THREAD

**Discussion by jchamlin** [shriprem/Goto-Line-Col-NPP-Plugin#36] [Q&A] [answered]
I'm working on file parsing and have a region of a file identified by a start column and start line number, and end column and end line number. And I would like that text to be selected when the file is loaded into Notepad++ by command-line. That would allow selecting of the block that caused a problem when clicking on an error, or jumping directly to some block of text to begin editing.

I would like to request a feature enhancement that would allow the ability to select the text between the Goto line/column position, and a new Select to line/column position. It could be a feature that could be enabled / disabled in Preferences, and if enabled would add two more line/column controls to the Goto Line, Column panel:

```
Goto:
  Line: 7
  Char column: 74
Select to:
  Line: 8
  Char column: 22
```

The cursor position would still remain at the Go to value, but the text between the two positions would be selected. If the cursor position and/or text that is selected changes, and the options are set to auto update the panel, the new cursor position would be updated and also the new selected region's end line/column would be updated too.

The -pluginMessage=GLC could be modified to add a new option, "s", which would specify a pair of numbers like 8,22 for the select to funcitonality.

Thank you in advance for your consideration.

**Comment by shriprem** [comment]
First of all, thank you very much for your enhancement idea.

I very much like the command line enhancement part of your suggestions. It seems to me that an implementation of just this command-line part will be sufficient to meet the goal stated at the start of your post.

Hence I am a bit less enthused with the panel interface enhancement part. Reasons:
1. The panel interface will get more cluttered and confusing to users. Although, this can be alleviated by a default hiding of these new fields from the panel, and making them visible only via a setting in the Preferences dialog. However, this alleviation measure will:
   a. Keep the new feature hidden and less discoverable to most users.
   b. And hiding/showing of the new columns, and the entailed up/down sliding of the 'Cursor Position Data' info box for a compact fit on the panel will require more implementation effort.

2. Notepad++ features a rich variety of selection features such as multiple selections and column selections. In contrast, the suggested new selection columns will only offer a single basic selection, and hence not much value addition.

I think I will be able to quickly implement the command line enhancement part of your suggestions. So, that is the good news. However, I will need more persuasive arguments from you for the panel enhancement part.

**Comment by jchamlin** [comment]
Thank you for responding so quickly. The command
-line option would be a great addition and provide the functionality I'm looking for.

However, the command-line is slow. Right now the command line launches a new NPP process, but that process communicates to the already running instance and then shuts down. And that's pretty slow.

So, to speed things up, I'm working on JNA access using Java to send messages to Notepad++. So, having all the controls available in the Goto Line, Column panel would allow me to interact with the panel through messages to change the value. So, unless there's a way to send the plugin a message directly through SendMessage(), the controls would allow interaction through the User32 API. Getting all the control IDs seems pretty easy: I just create a macro and look at the macro in shortcuts.xml.to get the values that are needed to be sent with SendMessage(). I haven't tried with the Goto Line, Column controls yet, but I assume I should be able to record a macro that interacts with the panel and get those IDs too.

**Reply by shriprem** [reply]
I now have a slightly better understanding, I believe, of what you are looking for. Have you considered using one of the scriptable plugins for Notepad++, such as NppExec, PythonScript or LuaScript?

I myself have both NppExec and PythonScript installed with my Notepad++. I highly recommend either one of them. I haven't used LuaScript plugin but I expect that it will have features comparable to the PythonScript plugin.

With any one of these plugins, all you need to do is:
1. Have that plugin's console open in NPP. This console will also be a Scintilla editor component running within the NPP shell.
2. Populate the console with a line of command text and an ENTER key code to execute that command.

This way, you will be able to access pretty much most of the message APIs of [Scintilla](https://www.scintilla.org/ScintillaDoc.html) and [Notepad++](https://npp-user-manual.org/docs/plugin-communication/). You will be able to select text easily, and do much much more. 

1. Both NppExec and PythonScript come with adequate help documentation.
2. There is a [user forum](https://sourceforge.net/p/npp-plugins/discussion/672146/) for NppExec.
3. If you need more support, there are very friendly folks at the [Notepad++ Community](https://community.notepad-plus-plus.org) board. In fact, I have seen script snippets for NppExec and PythonScript in some of the user posts here.

If you still need more assistance, feel free to reach out to me. Good luck!

But for now, I presume, I don't need to take any action on your enhancement requests for the plugin.

**Reply by jchamlin** [reply]
I think having the ability to launch notepad++.exe with the Goto Line, Column plugin installed and with command line arguments to highlight a region of text is a superior, and simpler approach to the general purpose need of clicking on a link and opening it in Notepad++ with the region selected. I was able to automate the process of opening the document and then selecting a region of text using a Regular Expression search. It, unfortunately, has the downside of littering the Find dialog with the regular expressions, and lots of Java code to achieve the effect. This could all easily be incorporated into the Goto Line, Column plugin since it has 95% of the code needed to achieve the task already written. I do think this feature enhancement is still the best approach to the overall problem.

Now, as promised, here are the details on what I have working:

The regex to highlight between a start line,column and end line,column has the following syntax:
`\A(?-s)(.*\R){startline-1}.{startcolumn-1}\K(.*\R){endline-startline}.{endline-startline == 0 ? endcolumn - startcolumn : endcolumn - 1}`

Example (select text between (313,295) and (314,17):
`\A(?-s)(.*\R){312}.{294}\K(.*\R){1}.{16}`

The Java code I'm using uses JNA to access the Windows API directly and was successfully able to perform the task. I'll share the code here just in case you'd like to see it. Ideally, it would be nice to be able to send a single message to the Goto Line, Column plugin to achieve the same results (and not litter the Find dialog's history).

```
	private static DesktopWindow hNPPWindow = null;
	private static String lastFile = null;
	private static boolean findDialogOpened = false;

	public static void highlightRegion(final String filename, final int startLine, final int startColumn, final int endLine, final int endColumn)
			throws IOException {
		// logger.debug("NotepadPlusPlus.highlightRegion() starting...");
		if (!filename.equals(lastFile)) {
			final String[] command = { "C:\\Program Files\\Notepad++\\notepad++.exe", "-n" + startLine, "-c" + startColumn, filename };
			Runtime.getRuntime().exec(command);
			lastFile = filename;
			try {
				Thread.sleep(1000);
			} catch (final InterruptedException e) {
				// ignore
			}
		}
		if (hNPPWindow == null) {
			hNPPWindow = Win32Utils.getWindowByFilePath("notepad++.exe");
		}
		if (hNPPWindow == null) {
			throw new RuntimeException("Notepad++ is not running");
		}
		final HWND hNppWnd = hNPPWindow.getHWND();
		if (!findDialogOpened) {
			user32.SendMessage(hNppWnd, NPPM_MENUCOMMAND, 0, IDM_SEARCH_FIND); // Open Find dialog
		}
		final HWND hFindDialog = user32.FindWindow("#32770", "Find"); // Find dialog window's handle
		// user32.EnumChildWindows(hFindDialog, new ListChildWindowsFunc(), null);
		final HWND hComboBox = user32.FindWindowEx(hFindDialog, null, "ComboBox", null);
		final HWND hTextEditBox = user32.FindWindowEx(hComboBox, null, "Edit", null);
		final String regex = computeRegex(startLine, startColumn, endLine, endColumn);
		user32.SendMessage(hTextEditBox, WM_SETTEXT, 0, regex); // populate search string
		final HWND hWrapTextCheckbox = user32.FindWindowEx(hFindDialog, null, "Button", "Wra&p around");
		user32.SendMessage(hWrapTextCheckbox, BM_SETCHECK, BST_CHECKED, 0);
		final HWND hRegularExpressionRadioButton = user32.FindWindowEx(hFindDialog, null, "Button", "Re&gular expression");
		user32.SendMessage(hRegularExpressionRadioButton, BM_CLICK, 0, 0);
		final HWND hFindNextButton = user32.FindWindowEx(hFindDialog, null, "Button", "Find Next");
		user32.SendMessage(hFindNextButton, BM_CLICK, 0, 0); // execute the search
		if (!findDialogOpened) {
			user32.PostMessage(hFindDialog, WM_CLOSE, 0, 0); // close the Find dialog
		}
		findDialogOpened = true;
		// logger.debug("NotepadPlusPlus.highlightRegion() done");
	}

	private static String computeRegex(final int startLine, final int startColumn, final int endLine, final int endColumn) {
		final int sl = startLine - 1;
		final int sc = startColumn - 1;
		final int el = endLine - startLine;
		final int ec = el == 0 ? endColumn - startColumn : endColumn - 1;
		final String result = "\\A(?-s)(.*\\R){" + sl + "}.{" + sc + "}\\K(.*\\R){" + el + "}.{" + ec + "}";
		return result;
	}
```

**Comment by shriprem** [comment, marked as answer]
I forgot to mention this earlier since it has been a few years:
I prototyped the core functionality for _GotoLineCol_ initially with the _PythonScript_ plugin, and only after that did I proceed to implement it in C++ as a standalone plugin. In fact, I later submitted those prototype scripts to be included as sample scripts in the _PythonScript_ plugin. If you have _PythonScript_ plugin installed within NPP, and navigate to NPP menu: `Plugins » Python Script » Scripts » Samples`, you will find the listings for **_GotoLineCol_** and **_GotoLineCol_Barebones_**. The scripts are located in the folder path: `Notepad++\plugins\PythonScript\scripts\Samples`. I recommend that you review the `GotoLineCol_Barebones.py` file. You will find how easy it is to achieve the cursor movement function of _GotoLineCol_ with _PythonScript_ as well.

To expedite your solution via _PythonScript_, I will also provide you the commands that you will need for automated text selection. Let's say you wish to select the text in a file from line: 313, col:195 to line: 414, col:16. To do this, you will need to execute the following two lines of commands from the PythonScript console box.
```
editor.gotoPos(editor.positionFromLine(313 - 1) + 195 - 1)
editor.setSelection(editor.positionFromLine(313 - 1) + 195 - 1, editor.positionFromLine(414 - 1) + 16 - 1)
```

#### Notes:
 1. The line and column numbers are zero-indexed in PythonScript. Hence you will need to decrement the line and column numbers by 1.
 2. The first command line is to bring the caret position of the file into the viewport of the NPP editor. Using the exactly equivalent line and col numbers may just bring the cursor to the bottom-right edge of the NPP viewport. Hence you may need to actually increment (or _decrement_ if repositioning after the viewport was closer to the bottom of the file) the line and col numbers by a little amount to bring the starting cursor position well within the viewport.

To automate this, you will need to populate each command line into the box highlighted in yellow in the clip below, and then execute a click of the _Run_ button highlighted in red.
![image](https://github.com/shriprem/Goto-Line-Col-NPP-Plugin/assets/45252729/33b40f8a-ecea-4a81-8959-93fbfaa8da5d)
